C/plugins/utils/extract_plugin_snapshot.cpp

```diff
@@ -0,0 +1,48 @@
+/*
+ * Utility to extract plugin snapshot tar archive
+ *
+ * Copyright (c) 2019 Dianomic Systems
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Amandeep Singh Arora
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <errno.h>
+
+extern int errno;
+
+/**
+ * Extract files from within plugin snapshot tar archive
+ *
+ *    Usage: extract_plugin_snapshot <plugin snapshot archive>
+ *
+ * @param argv[1]  relative/absolute path of plugin snapshot archive
+ */
+int main(int argc, char *argv[])
+{
+	if(argc < 2)
+	{
+		printf("Usage: %s <archive filename>\n", argv[0]);
+		return 1;
+	}
+
+        char *args[]={(char *) "/bin/tar", (char *) "-C", (char *) "/usr/local/foglamp/", (char *) "-xf", (char *) "PLACEHOLDER", NULL};
```

Comment:
It might not be /usr/local/foglamp. We should also cope with installations in different places that are reflect with FOGLAMP_ROOT environment variable.

Reply:
Yeah, considered that in discussions earlier today. The problem with reading FOGLAMP_ROOT environment variable in this program, is that someone could set this to say "/", set the tar.gz file contents with files to be overwritten in /usr/bin, /bin, or /sbin, etc and this setuid executable is available to modify contents of these critical directories as root user.

One possibility is to not allow executables in tar archive  (except perhaps links with 0777 permissions), or such alternate checks. Will think more about it.

Reply:
@MarkRiddoch - I added code to read FOGLAMP_ROOT env variable and verify foglamp installation at that path before proceeding with archive extraction. Please check.